<!doctype html>
<html>
<head>
	<title>Pinky example: Simulate fetching a random number</title>
	
	<!-- Google Code Prettify -->
	<script src="https://cdnjs.cloudflare.com/ajax/libs/prettify/188.0.0/prettify.js"></script>
	
	<style>
		body, textarea {
			font-family: Helvetica;
		}
		textarea {
			margin-left: 40px;
			font-size: 14px;
		}
		.str{color:#080}.kwd{color:#008}.com{color:#800}.typ{color:#606}.lit{color:#066}.pun{color:#660}.pln{color:#000}.tag{color:#008}.atn{color:#606}.atv{color:#080}.dec{color:#606}pre.prettyprint{padding:2px;border:1px solid #888}ol.linenums{margin-top:0;margin-bottom:0}li.L0,li.L1,li.L2,li.L3,li.L5,li.L6,li.L7,li.L8{list-style:none}li.L1,li.L3,li.L5,li.L7,li.L9{background:#eee}@media print{.str{color:#060}.kwd{color:#006;font-weight:bold}.com{color:#600;font-style:italic}.typ{color:#404;font-weight:bold}.lit{color:#044}.pun{color:#440}.pln{color:#000}.tag{color:#006;font-weight:bold}.atn{color:#404}.atv{color:#060}}
	</style>
	
	<!-- Include pinky.js -->
	<script src="../../pinky.js"></script>
	
	<!-- Our functions that use promises -->
	<script>
		/**
			Simulate fetching a random number from a remote source
			Time out if the pretend remote source responds too slowly
		*/
		function fetchRandom() {
			// Create a Pinky instance
			var pinky = new Pinky();

			// Let's pretend we're fetching the random number from a remote source
			// Randomly choose a "response time" between 0 and 100 milliseconds
			var responseTime = Math.random()*100;

			// We'll timeout if the source takes too long and reject the promise
			var timeoutTime = 50;
			var timeout = setTimeout(function() {
				pinky.reject(new Error('Request timed out'));
			}, timeoutTime);

			// Simulate an asynchronous fetch using setTimeout
			setTimeout(function() {
				// The response came back in time, clear our timeout
				clearTimeout(timeout);

				// Fulfill the promise with a random number when our pretend source responds
				pinky.fulfill(Math.random());
			}, responseTime);

			// Return the promise, which has a single method: then()
			return pinky.promise;
		}

		// A value is "valid" if it is not greater than 0.5
		function isValid(value) {
			return !(value > 0.5);
		}

		function doFetch(targetId) {
			var target = document.getElementById(targetId);
			
			// Call our asynchronous random number fetcher
			fetchRandom() // fetchRandom() returns a promise, which we'll call then() on below

			.then(function(value) {
					// In this onFulfilled handler, if the value is valid, we'll fulfill the next promise by returning the value
					// Or, if the value isn't valid, we'll reject the next promise by throwing an error
					if (!isValid(value)) {
						// Reject the next promise by throwing because the random number is too big
						var reason = new Error('Fetched value is invalid');
						reason.value = value;
						throw reason;
					}

					// Fulfill the next promise by returning if the value is small enough
					return {
						method: 'fetched',
						value: value
					};
				},
				null // Let the rejection reason bubble to the next promise's onRejected handler
			) // then() returns another promise

			.then(
				null, // Let the fulfillment value bubble to the next promise's onFulfilled handler

				function(reason) {
					// In this onRejected handler, if we return a value, the next promise will be fulfilled
					// Or, if we throw/re-throw, the next promise will be rejected

					// Fall back to generating a random number locally
					var newValue = Math.random();
					if (isValid(newValue)) {
						// Fulfill the next promise if our locally generated value is valid
						return {
							method: 'generated', 
							value: newValue
						};
					}

					// Throw a new error, including the message from the previously thrown error
					var newReason = new Error(reason.message+' and generated value is invalid');
					newReason.value = {
						fetched: reason.value,
						generated: newValue
					};
					throw newReason;
				}
			) // then() returns another promise

			.then(
				function(result) {
					target.innerHTML = 'Fulfilled: '+result.method+' value is valid: '+result.value;
				},
				function(reason) {
					target.innerHTML = 'Rejected: '+reason, reason.value;
				}
			);
		}
	</script>
</head>
<body onload="prettyPrint()">
	<h2>Pinky example: Simulate fetching a random number</h3>
	<p>This example has the following possible outcomes:</p>
	<ol style="list-style-type: decimal">
		<li><strong>Fulfilled</strong>: The fetched value is valid.</li>
		<li><strong>Rejected</strong>: The fetch timed out and the generated backup value is invalid.</li>
		<li><strong>Fulfilled</strong>: The fetch timed out or the fetched value is invalid, but the generated value is valid.</li>
		<li><strong>Rejected</strong>: The fetch timed out or the fetched value is invalid, and generated value is invalid.</li>
	</ol>
	
	<p>Click the button below to simulate an asynchronous fetch of a random value with error handling and validation.</p>

	<!--
		Note: Do not use the onclick attribute, it is bad practice. It is only used in this demo for brevity and to avoid including unrelated JavaScript code.
	-->
	<button onclick="doFetch('target');">Fetch number</button>
	
	<h3>Response:</h3>
	<textarea id="target" style="width: 600px; height: 40px"></textarea>
	
	<h3>Explanation:</h3>
	
	<h4>fetchRandom: a function that returns a promise</h3>
	<p>The <code>fetchRandom()</code> function simulates an asynchronous fetch and returns a promise that:</p>
	<ul>
		<li>Will be rejected if the "request" times out</li>
		<li>Will be fulfilled with the "response" if it comes back in time</li>
	</ul>
	
	<pre class="prettyprint">
/**
	Simulate fetching a random number from a remote source
	Time out if the pretend remote source responds too slowly
*/
function fetchRandom() {
	// Create a Pinky instance
	var pinky = new Pinky();

	// Let's pretend we're fetching the random number from a remote source
	// Randomly choose a "response time" between 0 and 100 milliseconds
	var responseTime = Math.random()*100;

	// We'll timeout if the source takes too long and reject the promise
	var timeoutTime = 50;
	var timeout = setTimeout(function() {
		pinky.reject(new Error('Request timed out'));
	}, timeoutTime);

	// Simulate an asynchronous fetch using setTimeout
	setTimeout(function() {
		// The response came back in time, clear our timeout
		clearTimeout(timeout);

		// Fulfill the promise with a random number when our pretend source responds
		pinky.fulfill(Math.random());
	}, responseTime);

	// Return the promise, which has a single method: then()
	return pinky.promise;
}
</pre>

	<h4>How to <em>then()</em>: using the promise returned by fetchRandom</h4>
	<p><code>fetchRandom()</code> is called and 3 chained <code>then()</code> calls are made, each of which returns a new promise:</p>
	<ol style="list-style-type: decimal">
		<li>With the first call to <code>then()</code> (on the promise returned by <code>fetchRandom()</code>), we'll pass the fulfillment value on if it's "valid" (not greater than 0.5), or let the timeout exception bubble to the next promise's onRejected handler.</li>
		<li>With the second <code>then()</code> call (on the promise returned by the first <code>then()</code> call), we'll let the fulfillment value bubble to the next promise's onFulfilled handler, or we'll handle invalid value exception by attempting to generate a valid value locally, returning it if we're successful, or throwing another exception if we're not.</li>
		<li>With the third <code>then()</code> call (on the promise returned by the second <code>then()</code> call), we'll print the fulfillment value and its source, or we'll print an error indicating we were unable to get a valid value.</li>
	</ol>
	<pre class="prettyprint">
// A value is "valid" if it is not greater than 0.5
function isValid(value) {
	return !(value > 0.5);
}

function doFetch(targetId) {
	var target = document.getElementById(targetId);
	
	// Call our asynchronous random number fetcher
	fetchRandom() // fetchRandom() returns a promise, which we'll call then() on below

	.then(function(value) {
			// In this onFulfilled handler, if the value is valid, we'll fulfill the next promise by returning the value
			// Or, if the value isn't valid, we'll reject the next promise by throwing an error
			if (!isValid(value)) {
				// Reject the next promise by throwing because the random number is too big
				var reason = new Error('Fetched value is invalid');
				reason.value = value;
				throw reason;
			}

			// Fulfill the next promise by returning if the value is small enough
			return {
				method: 'fetched',
				value: value
			};
		},
		null // Let the rejection reason bubble to the next promise's onRejected handler
	) // then() returns another promise

	.then(
		null, // Let the fulfillment value bubble to the next promise's onFulfilled handler

		function(reason) {
			// In this onRejected handler, if we return a value, the next promise will be fulfilled
			// Or, if we throw/re-throw, the next promise will be rejected

			// Fall back to generating a random number locally
			var newValue = Math.random();
			if (isValid(newValue)) {
				// Fulfill the next promise if our locally generated value is valid
				return {
					method: 'generated', 
					value: newValue
				};
			}

			// Throw a new error, including the message from the previously thrown error
			var newReason = new Error(reason.message+' and generated value is invalid');
			newReason.value = {
				fetched: reason.value,
				generated: newValue
			};
			throw newReason;
		}
	) // then() returns another promise

	.then(
		function(result) {
			target.innerHTML = 'Fulfilled: '+result.method+' value is valid: '+result.value;
		},
		function(reason) {
			target.innerHTML = 'Rejected: '+reason, reason.value;
		}
	);
}
</pre>
	
</body>
</html>
